client/network/src/service.rs

```diff
@@ -560,6 +568,71 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
+	/// given peer.
+	///
+	/// The returned `Future` finishes only after all the notifications have been queued or the
+	/// substream has been closed.
+	///
+	/// This operation is atomic in the sense that either all or none of the notifications have
+	/// been enqueued. In particular, it is guaranteed that no notifications have been transmitted
+	/// if the operation is interrupted by dropping the `Future`.
+	///
+	/// An error is returned if there exists no open notifications substream with that combination
+	/// of peer and protocol, or if the remote has asked to close the notifications substream.
+	///
+	/// If the remote requests to close the notifications substream, all notifications successfully
+	/// enqueued with this method will finish being sent out before the substream actually gets
+	/// closed, but attempting to enqueue more notifications will now return an error. It is also
+	/// possible for the entire connection to be abruptly closed, in which case enqueued
+	/// notifications will be lost.
+	///
+	/// The protocol must have been registered with `register_notifications_protocol` or
+	/// `NetworkConfiguration::notifications_protocols`.
+	///
+	/// # Usage
+	///
+	/// This method waits until there space is available in the buffer of messages towards the
+	/// given peer. If the peer processes notifications at a slower rate than we send them, this
+	/// buffer will quickly fill up.
+	///
+	/// As such, you should never do something like this:
+	///
+	/// ```ignore
+	/// // Do NOT do this
+	/// for peer in peers {
+	///     network.send_notifications(peer, ..., notifications).await;
+	/// }
+	/// ```
+	///
+	/// Doing so would slow down all peers to the rate of the slowest one. A malicious or
+	/// malfunctioning peer could intentionally process notifications at a very slow rate.
+	///
+	/// Instead, you are encouraged to maintain your own buffer of notifications on top of the one
+	/// maintained by `sc-network`, and use `send_notifications` to progressively send out
+	/// elements from your buffer. If this additional buffer is full (which will happen at some
+	/// point if the peer is too slow to process notifications), appropriate measures can be taken,
+	/// such as removing non-critical notifications from the buffer or disconnecting the peer
+	/// using [`NetworkService::disconnect_peer`].
+	///
+	///
+	///  Notifications              Per-peer buffer
+	///    broadcast    +------->   of notifications   +-->  `send_notifications`  +-->  Internet
+	///                     ^       (not covered by
+	///                     |         sc-network)
+	///                     +
+	///       Notifications should be dropped
+	///              if buffer is full
+	///
+	pub async fn send_notifications(
+		&self,
+		target: PeerId,
+		engine_id: ConsensusEngineId,
+		notifications: impl ExactSizeIterator<Item = impl Into<Vec<u8>>>,
+	) -> Result<(), SendNotificationError> {
```

Comment:
As an API suggestion, it might be nicer to return a concretely-typed future that has a method

```rust
/// Discard this `send_notifications` future, yielding the underlying buffer if it has not completed.
fn discard(self) -> Option<Self::Buffer>;
```

Reply:
An unrelated point: now with the `SendNotificationsError` we have effectively two ways to tell if a peer is disconnected. A guarantee that a disconnect event will always follow as well would be great because then we don't have to handle that case immediately but can just defer to the next iteration of our event-handling loop, which will reduce the amount of code-paths.

Reply:
:+1: we would need either of the following, correct?

1. Be able to check whether the buffer for a given peer has capacity before invoking `send_notifications` similar to [`Sink::poll_ready`](https://docs.rs/futures/0.2.1/futures/sink/trait.Sink.html#tymethod.poll_ready).

2. Be able to cancel the `Future` and get the passed `notifications` back as suggested in https://github.com/paritytech/substrate/pull/6591#discussion_r450918308.

This is especially relevant as one should not do `send_notifications(my_notifications).await`. Without either of the suggestions above one would need to do a possibly expensive `clone`, right?

```rust
let notification_clone = notification.clone();
match poll!(notifications.send_notifications(notification)) {
  Poll::Ready(Ok(())) => {},
  Poll::Ready(Err(e)) => handle_error(e),
  Poll::Pending => re_enqueue(notification_clone),
};
```

Or do we require the caller to pass a type that only clones the underlying data when `Into::into<Vec<u8>>` is actually called. In that case when the `send_notifications` future returns  `Poll::Pending` no `clone` of the actual data would occur.

Reply:
I've adjusted the PR.